### DISCUSSION TITLE
[help] Using tar_render() within tar_map()

### DISCUSSION THREAD

**Discussion by gls21** [ropensci/tarchetypes#175] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

# Reprex

## _targets.R file:

```
library(targets)
library(tarchetypes)
library(here)

### 

values <- data.frame(name = c("one", "two", "three"),
                     value = c(1, 2, 3))

mapped_targets <- tarchetypes::tar_map(
  
  values = values,
  
  tar_target(simple_calc, 10 + value),
  
  tar_render(report,
             path = here("report_template.Rmd"),
             output_file = here(paste0("results_", name, ".html")))
  
)
```

## report_template.Rmd file (in same directory):

```
---
title: "Untitled"
date: "`r Sys.Date()`"
output: html_document
---
```

```
{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(targets)

tar_read(simple_calc)
print(simple_calc)

```
![image](https://github.com/ropensci/tarchetypes/assets/91906620/072b4852-cba4-44e4-95e8-ced783878147)

# Desired result
I am trying to use tarchetypes::tar_map() to iterate over each row of the 'values' data frame, creating the defined targets for each. I am then trying to render an R markdown document - one for each of the rows of the 'values' data frame. Each file should be named accordingly - here I am trying to create 3 html files: "results_one.html", "results_two.html", and "results_three.html", where the names of the files are dynamically generated. In each of these Rmarkdown documents, I am trying to use tar_read(simple_calc) to get the result of the 'simple calculation' for each of the rows of the 'values' data frame. So in "results_one.html" it should print 11, in "results_two.html" it should print 12, and in "results_three.html" it should print 13. 

# Error message received when I run tar_make():
```
Error:
! Error running targets::tar_make()
Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
Debugging guide: https://books.ropensci.org/targets/debugging.html
How to ask for help: https://books.ropensci.org/targets/help.html
Last error message:
    object 'name' not found
```

# What have I investigated / tried?
Similar questions asked before here https://github.com/ropensci/tarchetypes/issues/18 and here https://github.com/ropensci/tarchetypes/discussions/104 though these don't seem to exactly match my issue? tar_render_raw() doesn't seem quite the thing I want?

I have also read in the tarchetypes documentation that "tar_map() is incompatible with tar_render() because the latter operates on preexisting tar_target() objects" so this suggests what I'm trying to achieve maybe isn't possible? If this is the case, is there a work around / alternative method to achieve the desired result?

# Session Info
```
> sessionInfo()
R version 4.4.0 (2024-04-24 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 10 x64 (build 19044)

Matrix products: default


locale:
[1] LC_COLLATE=English_United Kingdom.utf8 
[2] LC_CTYPE=English_United Kingdom.utf8   
[3] LC_MONETARY=English_United Kingdom.utf8
[4] LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.utf8    

time zone: Europe/London
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] here_1.0.1        tarchetypes_0.9.0 targets_1.7.0    

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5       cli_3.6.2         knitr_1.46        rlang_1.1.3      
 [5] xfun_0.43         processx_3.8.4    generics_0.1.3    data.table_1.15.4
 [9] glue_1.7.0        backports_1.4.1   rprojroot_2.0.4   ps_1.7.6         
[13] fansi_1.0.6       tibble_3.2.1      base64url_1.4     yaml_2.3.8       
[17] lifecycle_1.0.4   compiler_4.4.0    dplyr_1.1.4       codetools_0.2-20 
[21] igraph_2.0.3      fs_1.6.4          pkgconfig_2.0.3   rstudioapi_0.16.0
[25] R6_2.5.1          tidyselect_1.2.1  utf8_1.2.4        parallel_4.4.0   
[29] pillar_1.9.0      callr_3.7.6       magrittr_2.0.3    tools_4.4.0      
[33] withr_3.0.0       secretbase_0.5.0 
```

**Comment by wlandau** [comment]
It's the same underlying issue as #18 and #104, as explained in https://github.com/ropensci/tarchetypes/issues/18#issuecomment-718293603. You could consider a similar `purrr::map2()` workaround as in that post, or use `tar_eval()` for static branching as explained at https://books.ropensci.org/targets/static.html#metaprogramming

**Reply by gls21** [reply]
Thank you so much for getting back to me - I did read those responses, I just couldn't work out how to fit that in with my example? Would it be possible for you to provide the code show where and how those functions would fit into my example please?

**Reply by ginolhac** [reply]
I think this could help, inspired from @wlandau  examples:

`_targets.R`:

``` r
library(targets)
library(tarchetypes)
values <- list(name = rlang::syms(c("report1", "report2")), 
               file = c("report1.Rmd", "report2.Rmd"))
list(
  tar_target(params, data.frame(par = c("one", "two", "three"),
                                value = c(1, 2, 3))),
  tar_eval(
    tar_render(name, file), values = values
    )
)
```

`report1.Rmd`:

~~~ r
---
title: 'report.Rmd file'
output_format: html_document
---
Assume these lines are in a file called report1.Rmd.
```{r}
tar_read(params)
```
~~~

Running:
``` r
> tar_make()
▶ dispatched target params
● completed target params [0.001 seconds]
▶ dispatched target report2
● completed target report2 [0.443 seconds]
▶ dispatched target report1
● completed target report1 [0.351 seconds]
▶ ended pipeline [0.989 seconds]
```

Rendering of `report1.html`:
![image](https://github.com/ropensci/tarchetypes/assets/710181/71e0ae21-222e-4ffa-8ec3-28648bca6e17)

**Reply by gls21** [reply]
Hi, thanks for providing this. I've tried it out and it's still not quite doing what I'd like to achieve? I added a bit to your example to add in a target that selects a row at random from the params dataframe. This is to simulate me putting different content into each of the reports. 

```
library(targets)
library(tarchetypes)

values <- list(name = rlang::syms(c("report1", "report2", "report3", "report4", "report5")), 
               file = c("report1.Rmd", "report2.Rmd", "report3.Rmd", "report4.Rmd", "report5.Rmd"))

list(
  tar_target(params, data.frame(par = c("one", "two", "three", "four", "five"),
                                value = c(1, 2, 3, 4, 5))),
  tar_target(testing, params[sample(nrow(params), 1), ]),
  tar_eval(tar_render(name, file), values = values)
)
```

However, each report is being rendered with the same row, suggesting it's not randomly sampling a different row for each report? 
![image](https://github.com/ropensci/tarchetypes/assets/91906620/2ad5d831-fba8-4f6e-a59e-32b1c49a15d8)
![image](https://github.com/ropensci/tarchetypes/assets/91906620/bea14f64-4ef0-42de-87b0-cface3191f08)
![image](https://github.com/ropensci/tarchetypes/assets/91906620/a7853383-0c99-468a-bac0-872c81e50cd8)
etc...

Have you got any suggestions on how this could work please?

Thank you!

**Reply by ginolhac** [reply]
Make sense to me since in `testing` the rule is run once so all reports have the same seed, hence the same number.

But if the sampling is done inside the reports, I do obtain different values:

reports.Rmd contains:
~~~ r
---
title: 'report.Rmd file'
output_format: html_document
---
Assume these lines are in a file called report.Rmd.
```{r}
tar_load(params)
params[sample(nrow(params), 1), ]
```
~~~

render report1
![image](https://github.com/ropensci/tarchetypes/assets/710181/d16bfd5a-5eae-428b-b1ef-cf949f4ed347)

render report2
![image](https://github.com/ropensci/tarchetypes/assets/710181/3fe5d617-b36a-4231-90da-fd04ce0c6dfe)

**Comment by bart1** [comment]
For future reference I solved a similar problem by calling `rmarkdown::render` directly and using params:
``` 
values <- data.frame(name = c("one", "two", "three"),
                     value = c(1, 2, 3))

require(targets)
writeLines(
'---
title: "Untitled"
date: "`r Sys.Date()`"
output: html_document
params:
  test: "default"
---
```{r test}
params$test
```',"report.Rmd")

mapped_targets <- tarchetypes::tar_map(

  values = values,

  tar_target(simple_calc, 10 + value),

  tar_target(report,{
    rmarkdown::render("report.Rmd", output_file = tar_name(), params=list(test=simple_calc))
  })


)
list(mapped_targets)
```
